Reduce initial request size

freimair opened the pull request:
<!-- 
- make yourself familiar with the CONTRIBUTING.md if you have not already (https://github.com/bisq-network/bisq/blob/master/CONTRIBUTING.md)
- make sure you follow our [coding style guidelines][https://github.com/bisq-network/style/issues)
- pick a descriptive title
- provide some meaningful PR description below
- create the PR
- in case you receive a "Change request" and/or a NACK, please react within 30 days. If not, we will close your PR and it can not be up for compensation.
- After addressing the change request, __please re-request a review!__ Otherwise we might miss your PR as we tend to only look at pull requests tagged with a "review required".
-->

Fixes https://github.com/bisq-network/bisq/issues/2547, fixes https://github.com/bisq-network/bisq/issues/2549, fixes https://github.com/bisq-network/bisq/issues/3763, part of https://github.com/bisq-network/projects/issues/25

This PR reduces the initial request size to a O(1). By numbers, the number of historical keys to be sent to the seednodes is 1 now. With 1.3.4, that is 1 vs. 100k+ object keys and 2x1kB vs. 2x2500kB at startup.

## Details
This is a big step towards a more pleasant user experience and due to changes to parts of the data store implementation it is also high risk.
I see the split data store implementations as a temporary solution until I can proceed to https://github.com/bisq-network/projects/issues/29. Hence, even as there could be more tests, there aren't.

## Testing efforts
I created a number of test classes handling the basics of the migration and update scenarios, everyday use cases plus a few special cases. Additionally, there is a shell script running a kind of manual integration test which helped in finding more issues. The "manual" test takes care of setting up a testing testnet and through the steps upgrades pieces of software as well. Take a look at it if you so please. The results of the test have to manually observed although there is a `result.log` once finished which lists all the request and response sizes. (please not that some logs are put there twice because that is simply how it is).

Please give this a thorough test. If there are any questions, please feel free to contact me on keybase.

## EDIT:

as always when trying to test disk IO, travis complains and fails the tests. Please try the PR locally anyways. I will deactivate the tests for travis before the PR is merged.

freimair committed:
Initial tests

freimair committed:
Version helper uses real Bisq version

freimair committed:
Tests use static fixtures

freimair committed:
Migration test succeeds

freimair committed:
GetMap test succeeds

freimair committed:
GetMapSinceFilter works

freimair committed:
updateScenario works

freimair committed:
freshInstall scenario works

freimair committed:
put scenario works

freimair committed:
Introduce SplitStore and Service to handle BL

freimair committed:
Moved file handling logic to SplitStoreService

freimair committed:
Activate other split data stores

freimair committed:
Refactored tests

freimair committed:
Initial tests for requests

freimair committed:
Requests are smaller

freimair committed:
Prevent adding duplicate data

freimair committed:
Enforce version format and filter faulty ones

Instead of trying to find faulty keys, we stick
to the mechanics we have in place. This way,
we have less false positives.

However, we enforce a certain Bisq version
format, as parsing and business logic relies
on that now.

freimair committed:
Test incoming special keys from the future

freimair committed:
Javadoc cosmetics

freimair committed:
Init integration test script

freimair committed:
Persist live database after split

freimair committed:
Integration test script tests stock

freimair committed:
Use version history array

Turns out, scanning resources does not work reliable enough. Thus,
an array of Strings denoting historical Bisq versions it is.

An other way to put it is it is an array denoting which data stores
are there in the resources.

freimair committed:
Use full sync between seednodes

If we use the diff sync between seed nodes we create a race condition
where when a seednode gets updated to the new system, it does not
sync up properly with the other seed nodes. And that would be fatal.

So for the time being, when a seednode asks for data, it uses the
"old" big requests with all object keys. Should not be a problem
for now since they have enough bandwidth.

freimair committed:
Integration test script tests new system

freimair committed:
Cleanup integration test script

freimair committed:
Ignore some tests

CI does have troubles with tests which do file operations. Thus,
these tests have been disabled. They have been useful during
development and are useful for testing locally, though.

freimair committed:
Adjust P2PDataStorage tests

freimair committed:
Add another migration test

This test addresses the migration scenario where a user does not
upgrade on the first possible occation (to the first Bisq version
that has the new database structure in place) but does so later.

freimair committed:
Reactivate ignored tests

sqrrm commented on p2p/src/main/java/bisq/network/p2p/storage/persistence/SplitStoreService.java:
```diff
@@ -0,0 +1,190 @@
+            return null;
+
+        PersistableNetworkPayload previous = store.getMap().putIfAbsent(hash, payload);
+        persist();
```
Seems excessive to persist if nothing was written, but I notice that's what MapStoreService does.

freimair replied:
agreed. However, it is not trivial to find whether there has been something added to the map or not since the nested `put(...)` only returns the old value.

I believe that is why MapStoreService does this like so (and that might be one source of our IO issues).

I decided to stick with the "proven" way of doing things until we can address the whole disk IO issue in a more general way https://github.com/bisq-network/projects/issues/29

dmos62 replied:
Reading this is a bit difficult, because `getMap()` is easy to confuse with `store.getMap()`.

`getMap().containsKey(hash) == false` implies `store.getMap().containsKey(hash) == false`, because `getMap()` is a union of `store` and `store`s inside `history`, so L76 here will always insert a new entry. I'd like to point out that these `.put` and `.putIfAbsent` implementations (their returns) have different semantics than in `java.util.Map` (`.put` returns `void`). Same in `MapStoreService`.

sqrrm commented on p2p/src/main/java/bisq/network/p2p/storage/persistence/SplitStoreService.java:
```diff
@@ -0,0 +1,190 @@
+    public Map<P2PDataStorage.ByteArray, PersistableNetworkPayload> getMap() {
+        HashMap<P2PDataStorage.ByteArray, PersistableNetworkPayload> result = new HashMap<>();
+        result.putAll(store.getMap());
+        history.forEach((s, store) -> result.putAll(store.getMap()));
```
getMap() seems to be called a fair bit and it's recreating the map every time. Not sure how expensive that is, but perhaps some caching should be considered.

freimair replied:
I thought about that. The only thing that can be cached is loading the historical data stores into the result. And since these are read-only anyways, there is nothing that changes. So why not cache it?

Well, the maps of the stores are in memory as well. So the data is already there and ready to be transferred. The only redirection that happens when we do it like this is that we iterate through all stores and add the data to the result. Compare that to a "cached" map of historical data, the data has to be transferred as well. The only thing we gain by caching the data is we need more RAM while saving on a few instructions to iterate through history.

IMO The small computational overhead does outweigh adding the complexity of another caching layer that needs to be maintained.

Anyways, this piece of code is going to be deprecated by https://github.com/bisq-network/projects/issues/29 anyways. Then is the time IMO to really dial down performance.

sqrrm commented on p2p/src/main/java/bisq/network/p2p/storage/persistence/SplitStoreService.java:
```diff
@@ -0,0 +1,190 @@
+            filter = filter.replace("since ", "");
+            if (!filter.equals(Version.VERSION)) {
+                String finalFilter = filter;
+                history.entrySet().stream().filter(entry -> Integer.valueOf(entry.getKey().replace(".", "")) > Integer.valueOf(finalFilter.replace(".", ""))).forEach(entry -> result.putAll(entry.getValue().getMap()));
```
Try to keep lines < 120 chars

sqrrm commented on p2p/src/main/java/bisq/network/p2p/storage/persistence/SplitStoreService.java:
```diff
@@ -0,0 +1,190 @@
+    protected void readFromResources(String postFix) {
+        // check Version.VERSION and see if we have latest r/o data store file in working directory
+        if (!new File(absolutePathOfStorageDir + File.separator + getFileName() + "_" + Version.VERSION).exists())
+            makeFileFromResourceFile(postFix); // if we have the latest file, we are good, else do stuff // TODO are we?
```
This checks if the latest file exists...

sqrrm commented on p2p/src/main/java/bisq/network/p2p/storage/persistence/SplitStoreService.java:
```diff
@@ -0,0 +1,190 @@
+        else {
+            // load stores/storage
+            File dbDir = new File(absolutePathOfStorageDir);
+            List<File> resourceFiles = Arrays.asList(dbDir.list((dir, name) -> name.startsWith(getFileName() + "_"))).stream().map(s -> new File(s)).collect(Collectors.toList());
```
... if it does exist we make a list of all the files we have.

Also, line breaking streams make them a lot easier to follow.

sqrrm commented on p2p/src/main/java/bisq/network/p2p/storage/persistence/SplitStoreService.java:
```diff
@@ -0,0 +1,190 @@
+            store = readStore(getFileName());
+            resourceFiles.forEach(file -> {
+                SplitStore tmp = readStore(file.getName().replace(postFix, ""));
+                history.put(file.getName().replace(postFix, "").replace(getFileName(), "").replace("_", ""), tmp);
```
Then we add them to our history. What is the key, not easy to follow. Do we depend on having all the previous stores since we're only checking for the last one?

freimair replied:
> Do we depend on having all the previous stores since we're only checking for the last one?

exactly. And you may have spotted something here.

When we have the last one, we already did the transformation to the new split store system. Hence, I assumed that from now on we already have the historical stores. On a Bisq update, only the newest store is checked and if it isn't there, we need to copy it from resources and un-duplicate things in our live data store.

But you are correct, that might need a bit more effort to make the whole system more robust.

freimair replied:
just added some additional logic to address this issue.

I could not stay true to the rest of the coding style and did my own design. And by doing that, I saved on 30 loc.

sqrrm commented on p2p/src/main/java/bisq/network/p2p/storage/persistence/SplitStoreService.java:
```diff
@@ -0,0 +1,190 @@
+        List<String> resourceFiles = versions.stream().map(s -> getFileName() + "_" + s + postFix).collect(Collectors.toList());
+
+        // if not, copy and split
+        resourceFiles.forEach(file -> {
```
`file` is a filename string...

sqrrm commented on p2p/src/main/java/bisq/network/p2p/storage/persistence/SplitStoreService.java:
```diff
@@ -0,0 +1,190 @@
+        // if not, copy and split
+        resourceFiles.forEach(file -> {
+            final File destinationFile = new File(Paths.get(absolutePathOfStorageDir, file.replace(postFix, "")).toString());
+            final String resourceFileName = file;
```
...making this var a bit excessive.

sqrrm commented on p2p/src/main/java/bisq/network/p2p/storage/persistence/SplitStoreService.java:
```diff
@@ -0,0 +1,190 @@
+        store = readStore(getFileName());
+        resourceFiles.forEach(file -> {
+            SplitStore tmp = readStore(file.replace(postFix, ""));
+            history.put(file.replace(postFix, "").replace(getFileName(), "").replace("_", ""), tmp);
```
A helper method to do these string manipulations would help readability.

sqrrm commented on p2p/src/main/java/bisq/network/p2p/storage/persistence/SplitStoreService.java:
```diff
@@ -0,0 +1,190 @@
+            SplitStore tmp = readStore(file.replace(postFix, ""));
+            history.put(file.replace(postFix, "").replace(getFileName(), "").replace("_", ""), tmp);
+            // - subtract all that is in resource files
+            store.getMap().keySet().removeAll(tmp.getMap().keySet());
```
With this change I think we're making the resource files the final arbiter of what is the distributed data. Is that what we want? Even for full nodes that are always on.

From what I understood so far, nodes will keep getting data from the network. With new versions all new data since the last version will be stored in a new `resource_x.x.x_MAINNET` file and all nodes will restore this data from resources. It looks like it will just throw away its own local data on new versions though, that seems incorrect.

freimair replied:
> With this change I think we're making the resource files the final arbiter of what is the distributed data. Is that what we want? Even for full nodes that are always on.

This seems to be the only way of introducing "checkpoints" so that we do not have to query all the data.

> It looks like it will just throw away its own local data on new versions though, that seems incorrect.

It does not throw away any data. It only de-duplicates the data. If we have a data store "incoming" from resources, we remove the data that is contained in the resource store from our live data store.

An example:
Given _ResourceStore={1,3}_ and _LiveStore(t)={1,2,3,4,5}_ and we enter the copy and split business logic, the outcome would be having the _ResourceStore_ being copied to the working directory and the new _LiveStore(t+1)=LiveStore(t) \ ResourceStore = {2,4,5}_. Together, they hold {1,2,3,4,5}.

Bringing it all together:
In terms of initial data request size, we can now ask for all data except _ResourceStore_ simply by telling other that we have the "ResourceStore_1.3.4" -> sending {"1.3.4"} instead of {1,3}. And because in real live, the data stores count >100k objects, we save a whole lot of data. Additionally, we actually ask for all data "since" (instead of "except") and thus keep the request object count at O(1) in the future.

Again, and hopefully, the file operations will soon be superseded by https://github.com/bisq-network/projects/issues/29.

sqrrm commented on p2p/src/main/java/bisq/network/p2p/storage/P2PDataStorage.java:
```diff
@@ -205,24 +213,47 @@ public GetUpdatedDataRequest buildGetUpdatedDataRequest(NodeAddress senderNodeAd
+        if (result.isPresent())
+            return result.get();
+        else
+            return "";
```
Easier to read would be `        return result.orElse("");`
or even
```
        return collection.stream()
                .map(byteArray -> new String(byteArray.bytes).trim())
                .filter(s -> s.matches("^[0-9]\\.[0-9]\\.[0-9]$"))
                .findFirst()
                .orElse("");
```

sqrrm commented on p2p/src/main/java/bisq/network/p2p/storage/P2PDataStorage.java:
```diff
@@ -205,24 +213,47 @@ public GetUpdatedDataRequest buildGetUpdatedDataRequest(NodeAddress senderNodeAd
+    }
+
+    private String containsSpecialKey(Set<P2PDataStorage.ByteArray> collection) {
+        Optional<String> result = collection.stream().map(byteArray -> new String(byteArray.bytes).trim()).filter(s -> s.matches("^[0-9]\\.[0-9]\\.[0-9]$")).findFirst();
```
Is String(byte[]) safe enough? I suspect so but I'm not sure.

freimair replied:
at least I haven't found an issue yet

sqrrm replied:
I ask since it might act differently on different platforms. I think it would never be an issue for version values, but perhaps other values could cause problems, deterministic is nice.

freimair replied:
Just played around with it a bit. Doing an old-fashioned for-switch-if routine is about 10 times faster. However, 10 times faster means it takes an average of 2,6ms vs. 23,4ms (60k objects, average over 100 tries) for an operation that is performed twice during the runtime of Bisq (file loading time is about 3 seconds). In contrast, it takes 20 LOC more to achieve this. 

ad deterministic behavior: I haven't found any hint on non-deterministic behavior. Creating a String from byte[] is pretty standard. I, however, added a charset for good measure. Although "\d.\d.\d" contains nothing another charset would handle differently.

sqrrm commented on p2p/src/main/java/bisq/network/p2p/storage/P2PDataStorage.java:
```diff
@@ -205,24 +213,47 @@ public GetUpdatedDataRequest buildGetUpdatedDataRequest(NodeAddress senderNodeAd
-                .map(e -> e.bytes)
-                .collect(Collectors.toSet());
+        Set<byte[]> excludedKeys;
+        if (seedNodeRepository != null && seedNodeRepository.isSeedNode(networkNode.getNodeAddress()))
```
Braces around multi-line blocks really helps readability in my opinion.

sqrrm commented on p2p/src/main/java/bisq/network/p2p/storage/P2PDataStorage.java:
```diff
@@ -268,9 +299,18 @@ public GetDataResponse buildGetDataResponse(
+            tmp = this.appendOnlyDataStoreService.getMap();
+        else {
+            tmp = this.appendOnlyDataStoreService.getMap("since " + specialKey);
+        }
```
Is this the separation of old and new requests for data? More comments would be helpful for understanding.

Variable name `tmp` isn't really descriptive, better naming would help.

freimair replied:
> Is this the separation of old and new requests for data?

Correct. If there is a special key incoming we prefilter the data, if not, we do what we did before.

sqrrm reviewed (commented):
I've read through the core parts of the PR. For now just adding comments to gain a better understanding. This PR makes me concerned that we might lose data during upgrades or otherwise, but I don't have a full understanding yet.

freimair committed:
Refactoring and cleaning up

freimair commented:
just rebased to master and folded in the first batch of feedback. (Thanks @sqrrm for your efforts)

Build is failing because JDK12 does not support removing final-modifiers through reflections. If I remove the tests that rely on that everything would be green. Not sure how to handle that. And no, there is no other way to test this except loosening something that clearly is final to being not final in the first place.

freimair committed:
Cleanup

freimair committed:
Refactored test helpers

freimair committed:
Fixed a potential issue when loading stores

Previously, we only checked if the newest store is there. If it is
not, we copy and reload everything. If it is there, we are satisfied.

Now, we check for each and every store we should have and copy and
load them individually.

sqrrm commented on p2p/src/main/java/bisq/network/p2p/storage/persistence/SplitStoreService.java:
```diff
@@ -121,39 +119,28 @@ private int parseSpecialKey(String specialKey) {
+
+        // go through the list one by one
+        versions.forEach(version -> {
+            String filename = getFileName() + "_" + version + postFix; // postFix has a preceding "_"
```
There aren't any postfixes in the db dir, should the postFix be added here?

freimair replied:
oh my. sorry. bug. fix is coming up.

sqrrm commented on p2p/src/main/java/bisq/network/p2p/storage/persistence/SplitStoreService.java:
```diff
@@ -121,39 +119,28 @@ private int parseSpecialKey(String specialKey) {
+                history.put(version, readStore(getFileName() + "_" + version));
+            } else {
+                // either copy and split
+                history.put(version, copyAndSplit(version, postFix));
```
This will be done for all historical versions but copyAndSplit doesn't know about versions so it will make an identical copy of the resource data and put a version number on it.

I think there is something not right with this. We should be shipping this version with a split store (which will just be one old version resource). As time goes we will add more versioned resource files. It might be the old storage format doesn't have to ported in code at all. If we just rename the resource files we have to name_1.3.4_BTC_MAINNET this version is already good to go.

freimair replied:
> make an identical copy of the resource data and put a version number on it.

I believe you got it already. We rename the resource files and stick a "version" to it manually. So eg. with 1.3.6 we ship 3 stores: name_1.3.4_BTC_MAINNET, name_1.3.5_BTC_MAINNET, name_1.3.6_BTC_MAINNET. Then, as copyAndSplit knows about the version, each file is copied if necessary.

- the storage format remains the same. It is still Protobuf and files.
- `name_1.3.4_BTC_MAINNET` yes, exactly (I will be renaming the files in the PR once all is said and done)
- but in the future, there is not one big data store, but multiple - one for each release of Bisq
- why multiple?
  - because then we get some time relations and thus, chronological order. Once we have some order, we can start deciding which parts we keep in RAM (not in this PR, but down the line)
  - reduce the amount of data added to the repository by each release (deprecated by https://github.com/bisq-network/bisq/pull/4114)
  - in the long run, keep small chunks if possible

sqrrm commented on p2p/src/main/java/bisq/network/p2p/storage/P2PDataStorage.java:
```diff
@@ -213,19 +213,33 @@ public GetUpdatedDataRequest buildGetUpdatedDataRequest(NodeAddress senderNodeAd
+                .map(byteArray -> new String(byteArray.bytes).trim())
+                .filter(s -> s.matches("^[0-9]\\.[0-9]\\.[0-9]$"))
+                .findFirst()
+                .orElseThrow();
```
Please don't use exceptions as part of the expected execution path. This is not how Bisq uses exceptions so it causes confusion.

sqrrm reviewed (commented):
Thinking a bit about it I guess there is conceptually the same model we have now with one store, just split in many.

I am having a hard time getting a good overview of what's going on, but I think there is some muddling of concerns. A new client would not have to know about the old format as I see it. Only those nodes delivering data (seednodes typically) need to separate the kind of request they're getting and deliver data accordingly.

freimair committed:
Fixed wrong filename

freimair committed:
Refactoring

freimair committed:
Make special key platform-independent

freimair committed:
Fixed persisting wrong data store

freimair committed:
Renamed data stores

dmos62 commented on p2p/src/main/java/bisq/network/p2p/storage/P2PDataStorage.java:
```diff
@@ -205,24 +214,61 @@ public GetUpdatedDataRequest buildGetUpdatedDataRequest(NodeAddress senderNodeAd
-                .map(e -> e.bytes)
-                .collect(Collectors.toSet());
+        Set<byte[]> excludedKeys;
+        if (seedNodeRepository != null && seedNodeRepository.isSeedNode(networkNode.getNodeAddress())) {
```
Meaning of this predicate not immediately obvious: maybe `var weAreASeedNode = seedNodeRepository != null && seedNodeRepository.isSeedNode(networkNode.getNodeAddress())`?

dmos62 commented on p2p/src/main/java/bisq/network/p2p/storage/P2PDataStorage.java:
```diff
@@ -205,24 +214,61 @@ public GetUpdatedDataRequest buildGetUpdatedDataRequest(NodeAddress senderNodeAd
         Set<byte[]> excludedKeysFromPersistedEntryMap = this.map.keySet()
                 .stream()
                 .map(e -> e.bytes)
                 .collect(Collectors.toSet());
```
`.keySet().stream().map(e -> e.bytes).collect(Collectors.toSet())` is applied 3 times in this method. Moving it to a method (like `getKeySetInBytes(Map)`) would improve readability.

dmos62 commented on p2p/src/main/java/bisq/network/p2p/storage/P2PDataStorage.java:
```diff
@@ -205,24 +214,61 @@ public GetUpdatedDataRequest buildGetUpdatedDataRequest(NodeAddress senderNodeAd
     }
 
+    /**
+     * Create the special key. <br><br>
```
"Special key" is a problem semantically, in that it doesn't describe how it is special. Better to be concrete: it's the release version string in bytes. That's not my main point here, but I noticed a few phrasings like this so felt that it should be mentioned.

The actual problem is that there's string and byte manipulations in various files in this PR. You can put all this logic in `bisq/common/app/Version.java` under `toP2PDataStorageKey`, `static fromP2PDataStorageKey`, etc. Most places where strings are being passed around, `Version`s should be passed around instead. For example, `Version.history` would do better as `List<Version>`, than `List<String>`.

freimair replied:
I somehow can agree on 'special key' to not be ideal. Fix coming up...

On the other hand, I fully agree on manipulating Version strings is downright bad and not desirable. However, refactoring `version` format would affect big portions of Bisq. Thus, it would directly violate our doctrine of keeping PR small and right to the point. One has to draw the line somewhere, here it is.

dmos62 commented on p2p/src/main/java/bisq/network/p2p/storage/persistence/AppendOnlyDataStoreService.java:
```diff
@@ -55,6 +55,18 @@ public void readFromResources(String postFix) {
     }
 
+    /**
+     * TODO to be implemented
```
Probably forgot to remove this.

dmos62 commented on p2p/src/main/java/bisq/network/p2p/storage/persistence/SplitStoreService.java:
```diff
@@ -0,0 +1,201 @@
+ */
+@Slf4j
+public abstract class SplitStoreService<T extends SplitStore> extends MapStoreService<T, PersistableNetworkPayload> {
+    protected HashMap<String, SplitStore> history;
```
`HashMap<Version, SplitStore>` would be better.

freimair replied:
again, not touching the `version` format in order to keep the PR small and down to the point.

dmos62 commented on p2p/src/main/java/bisq/network/p2p/storage/persistence/SplitStoreService.java:
```diff
@@ -0,0 +1,201 @@
+        }
+
+        return result;
+    }
```
`getMap(String)` is cryptic. `getObjectsSince(Version)` would be much better. Notice that gets rid of the language/DSL. Doing a language is not necessary, because currently there's only one usage for these version keys here: getting objects from newer releases than the one specified.

`parseSpecialKey(finalFilter)` is called a lot, could be cached.

freimair replied:
you are correct, YAGNI right now. However, that will change swiftly once we get to reducing the memory footprint (which will start with https://github.com/bisq-network/projects/issues/29)

dmos62 commented on p2p/src/main/java/bisq/network/p2p/storage/persistence/SplitStoreService.java:
```diff
@@ -0,0 +1,201 @@
+        storage.queueUpForSave();
+
+        return historicalStore;
+    }
```
Why are we copying historical datastore checkpoints/splits out of resources? Because that's what the live-store-only solution we currently use does? For troubleshooting purposes? Seems unnecessary, though I may be missing something. I presume we keep the live store in the working directory, because we can't mutate resource files (is that even true?).

freimair replied:
> Why are we copying historical datastore checkpoints/splits out of resources? Because that's what the live-store-only solution we currently use does? For troubleshooting purposes? Seems unnecessary, though I may be missing something. 

the existing file storage mechanisms are all tailored to working in the working directory. Changing that would be equivalent to replacing the file storage mechanisms with something else entirely. Hence, leaving it as is (until https://github.com/bisq-network/projects/issues/29).

> I presume we keep the live store in the working directory, because we can't mutate resource files (is that even true?).

Well, obviously, the "live store" is something that is created and maintained by the Bisq app per installation and even per (configurable) app name and path. Placing that in the resources would only allow for a single instance of the data store AND would be overwritten and therefore invalidated on every app upgrade.

dmos62 reviewed (changes requested):
Requested changes. Mostly related to clean up, but still significant. I found readability difficult. Would appreciate more helpful names and more intermediary variables.

I wrote up a description of the PR and the problem it's solving. I'll include it below. Some of the things there may not be correct; it represents my ongoing learning about Bisq's P2P (and this PR). @freimair, after corrections, if they're necessary, would you consider including some of this information in an appropriate place in the code? I think future devs will appreciate an explanation of why this datastore design was chosen. I would appreciate feedback if I got something wrong.

------------------------

A datastore is a store of Bisq p2p objects. It is synced between peers-peers and peers-seednodes, 
and this way makes up a distributed database.

I presume that a peer's datastore is synced in this way:
  - a peer/client starts up;
  - it starts a "bulk sync" procedure:
    - gathers up UIDs (or hashes) of the objects it already has in its datastore;
    - uploads those UIDs to a seed node;
    - seed node takes a set difference of the passed UIDs and the UIDs of the objects it has;
    - seed node uploads to the peer/client the objects corresponding to that difference;
  - peer then enters p2p mode, listening for and publishing updates.

The problem is that this "bulk sync" procedure includes the peer sending a large number of UIDS
to the seed node, which can timeout, in case of insufficient network throughput, causing the
syncing to unrecoverably terminate.

That shows two problems:
  - Bisq does not recover from throughput issues;
  - the bulk sync is bandwidth-intensive.

This spec addresses the bandwidth-intensity.

Chronological information in the p2p objects would allow querying it by comparing objects to a
timestamp-like value; however, the p2p objects do not contain chronological information. It is
unclear whether introducing a timestamp would have any negative effects, but it would involve
making sensitive changes.

Freimair's proposal is about using chronologically-indexed datastore "checkpoints" that are 
distributed with releases
  - (and thus identical across clients and seednodes)
  - each checkpoint contains objects generated since the last checkpoint
    - the first checkpoint contains all prior objects
    - this effectively assigns chronological information to objects
      - precision of which, for the purposes of bulk sync, corresponds to frequency of releases;
  - this way a bulk sync can be initiated by identifying
    - the latest checkpoint that the client has
    - and the objects it has stored since then (objects not in any checkpoint known to him)
    - which reduces the bandwidth requirement
      - to the equivalent of uploading UIDs that are newer than the release used
        - because identifying the checkpoint is negligible bandwidth-wise.

Requirements for the checkpoint datastore:
  - initializable from the previous monolithic datastore or from scratch;
  - a "running" checkpoint has to be maintained;
    - which holds objects generated since the last "release" checkpoint;
  - at no point can the datastore be mutated, except via appends;
  - "release" checkpoints must be identical across peers and seednodes.

freimair commented:
@dmos62 thanks for your review. Appreciate it.

First of all, your PR description is correct.

Furthermore, I believe I can address all of your comments in summary:
I see this PR as an as-small-as-possible fix for the 3/4 problem. Yes, files are copied around where they probably do not need to be, APIs could be more readable, return values are ignored or misused, and a lot of other stuff. Thing is, most of the issues you raised have been in the code already and I do not want to touch them. They have proven themselves. Keep the changes small, keep the isolated. Its a minefield.

That being said, https://github.com/bisq-network/projects/issues/29 is in the queue. That project will get rid of the copying files around because there will be no more individual files, the API will need a major overhaul, the filter might get enhanced, the `isSeednode` code will be gone because it prevents our network from falling apart while upgrading from single data stores (now) to multiple (then) -After that is done, it is no longer needed.

Now to your obvious question: why not do it properly? I rather not spend time on reworking the same thing twice :smile: because I am a) a lazy bastard, and b) we need to use the resources we have efficiently. And of course, coding styles and our brains differ, so there is no perfect PR anyways.

freimair committed:
Replace TODO comment

dmos62 commented:
@freimair followed up above post with a Keybase message and we had a brief exchange there. I'll summarize how I see this situation and what my concerns are:

- P2P datastore code has a fair amount of technical debt, which makes making changes to it without compromises slow;
- we want to introduce changes quickly, which encourages us to make compromises;
- compromising changes tend to degrade reliability and are more likely to introduce mistakes;
- compromising changes require more resources to maintain: freimair points out that there are plans to replace large portions of this code entirely, which means that if all goes well those portions will not require maintenance;
- compromising changes require more resources to review.

My concerns:

- I've spent a long time on the PR (~2.5 days, I realise this says as much about my reviewing ability as it does about the PR), but I'm still not confident that all edge cases are covered; I don't feel comfortable seeing this merged at this time;
- I think that reliability, maintainability and reviewability are preferable to speed, especially in sensitive code, though speed has its importance too; it's a balancing act;
- introducing changes that are difficult to maintain, because if all goes well they won't have to be maintained, is planning for the best outcome; I don't think that's prudent;
- I think it's good to address this human element: PRs that make hasteful changes, or side-step technical debt, often require the reviewer to address that complexity anyway, so that he can make sure that the changes are correct, which can be frustrating for the reviewer (who usually always has his plate full as is).

Idea:

Stack a backup system on top of this in such a way that the mutations we perform cannot lose data. Of course we have to be aware that this is addressing complexity with more complexity. Maybe this is an example of how _not_ to solve this situation.

freimair committed:
Fixed a timing issue

chimp1984 commented:
A few "high level" comments/thoughts:

By splitting up the data into groups we are creating "blocks" (I use that terminology with intention). How to uniquely identify a block in a distributed system is a challenge. 

The approach used here makes the shortcut via the release maintainer to be the single entity to define what is the right "block". This might not be an issue (or at least does not make things worse as it is) but I am not 100% sure. The checkpoint terminology reveals the centralisation issue.

Is there another way to do it?
We have date fields in most (all?) of the relevant data structures and could use that for sorting the items but there could be multiple items with the same date. To get unique IDs ans sorting we could add the hash fields and create a tuple {date, hash} as ID  and sort primarily by date, secondarily by hash. 
But there can be a late insertion as well which could mutate our data blocks. For certain data we don't allow insertion outside a defined time period tolerance window (`DateTolerantPayload`). So we could consider blocks of a certain age as immutable at least. Newer blocks might get "reorged". Handling "reorgs" sounds like looking for toubles though...

Another approach could be to use a fixed time period window (e.g. 10 days) and fill up all data whose date falls into that period. The GetData request would send the index of the time window and the hash of its content. If the Seed node has a different hash it will send the whole package. For older data that should be very rare and the additional network overhead might be a good trade off with the more simple approach and low processing costs. The hash gets calculated once and is stored with the package. A problem is how to handle the case if a user has more data items than the seed node. The it would request each time again the package. If a seed is missing data it would get dossed by the network....

Another approach might be to use Bloom filters for the data request, but that is likely expensive for the seed node to process. A probably bigger problem than the 2.5 MB request is the load on the seed node to process the response.

dmos62 commented:
The main problem that we're trying to get around is that p2p objects are meant to be indexed by hash and don't have properties that would allow them to be indexable by something lightweight (like time). Bloom filters are normally the solution for these situations, but their processing is memory intensive, which opens up Bisq to denial-of-service attacks. It might be worthwhile to look at academic literature on the topic.

I can't dig deeper now, but I found some interesting articles following these links. Proabilistic set membership data structures are actively researched:
- https://scholar.google.com/scholar?q=bloom+filters+alternative+memory
- https://scholar.google.com/scholar?q=bloom+filters+memory

chimp1984 commented:
@dmos62 Thanks for your input!

After a discussion with @sqrrm we agreed the best way forward is to limit a first mitiagtion to trade statistic objects only. This will reduce bandwith requirement by 50% as well as seed node processing load by 50% (as its about 50% of the hashes). It will also reduce required RAM for the Bisq app if it is only loaded on demand.

Rough idea:

Prune existing trade statistics by removing all data older than a cut off date - e.g. 2 months (TBD). Beside that handling of trade statistics data is exactly as it is now but instead of 4 years of data we only deal with 2 month. I assume that will be about 1-5% of current data.
At each release we add a snapshot named with the release version which contains all historical data until a defined cutoff date (can have an overlap with the in-memory data to be on the safe side - details have to be defined once we work on it).
This historical data is available for new users as its in the binary. For not updated users they will request it with a new request call from the seed nodes. They pass the versions they have and receive the missing ones. This will be only done once the data is actually needed (user goes on historical trade statistics UI). We could even spit that UI in recent trades which covers the in-memory data and historical data which makes the request and/or loads the shipped resource files. After leaving this UI we unload that data from RAM. Loaded data and in-memory data might overlap but as its hashmaps those will get canceld out. 

This approach reduces a lot the risks as if anything goes wrong we only screw up trade statistics which are non-critical data.
If the concept of using version based buckets works out well it can be extended to the next data type like account witness objects (here we dont have the concept of on-deman data as all data are required).
 
Backward  compatibility:
We need to take care that not updated nodes still work properly. I guess there are no big issues beside potential extra load when a updated seed node request data from a not updated seed. Here we get a bit of extra load as the not updated seed node would deliver all the missing data (up to the hard cap limit). The updated node would ignore those historical data as they will not fit the date check predicate. With customizing the update of seeds by assigning a seed to connect with (one which has alreayd updated) at startup we can minimize that load. Beside that we could use other tools like capabilities, verion numbers, new data fields as flags, activation date to make the transition smooth.
Users who have not updated longer than 2 months would see gaps in the trade statistics, but that is not a major issue as that data is only used for informational purpose. We can even require an update via the filter after a certain time.

I might work on that myself in a couple of weeks if I find time. Otherwise if anyone wants to pick it up, please leave a message.

sqrrm commented:
Having discussed it and considering the risks it seems much safer to only focus on the tradestats to begin with. The risks are very high if any of the DaoState data or account witness data is lost. Losing some tradestats data is not critical in the same way but could still yield results. It would also help to prove the concept.

chimp1984 commented:
I started working on it: https://github.com/bisq-network/bisq/pull/4405 Lets move discussion over there....

freimair committed:
Added 1.3.6 and 1.3.7 data stores

freimair committed:
Update history version list

freimair committed:
Removed deprecated test mechanisms

Up until now, the tests needed to create a version history set for testing.
These changes had to be made to a final static variable during runtime.
Now, enough versions are in the history set anyways, so we could remove
this part of the testing framework.

freimair committed:
Cleanup imports

freimair commented:
A bit late, I know, but I can be online again, local infrastructure is fixed, and I take my time to comment on your discussions. Although I would suggest a call sometime, just to get somewhere.

### ad rough idea
pruning data does not solve the problem.
- it lowers usability
- we fix a piece of code that is at its limits with throwing even more code on it, thus increasing complexity instead of decreasing it
- it adds a lot of complexity down the road as we have to handle missing data throughout Bisq
- once Bisq gets more successful, we need to start pruning data a lot more rigorously (one month, 2 weeks?)

### ad another way to do it
- we cannot use time (as you guys already realized)
- the only thing we control is releases. someone throws a number of messages into a bucket and put a stamp on it. Then we ship this bucket. No issues with later additions, not even incomplete buckets are an issue because the network keeps the databases synced anyways.
- `The checkpoint terminology reveals the centralisation issue.` That is correct in a couple of ways:
  - `reveals`, because the issue is there already, cannot be discussed away
  - we have to deal with it eventually
  - by leaving it untouched, we do not move and cannot deal with "it"
  - we cannot solve all and everything now

### ad backwards compatibility
- seed nodes communicate using the old mechanics. nothing changed there. Because of all the reasons you mentioned and more.

### ad `if any of the DaoState data is lost`
- DaoState data isn't touched by this PR at all, and although the massive data store causes real issues, it does not have anything to do with the 3/4 problem

### ad loosing any data in general
- no data is deleted, no pruning, no shuffling around.
- the only thing that happens is to move data from the live database once a new bucket arrives (through app update) and thus, reduce the Disk IO load

### ad `Having discussed it and considering the risks it seems much safer to only focus on the tradestats to begin with. `
- focusing on tradestats is possible if you guys so desire
- however, I do not understand why there is this fear of loosing data because of this PR. Yet, pruning data - ie. intentionally loosing data, seems to be a way forward?
- I do however, fully understand, and have seen it happen, that we loose data once in a while because the current implementation cannot cope with the sheer amount of data anymore (because eg. it takes too long to write to disk and files get clipped [which by the way is addressed by this PR as well, at least for non-dao stuff])
- I do fully understand that we deny users access to Bisq which do not own a powerful dev machine
- I do fully understand that we deny users access to Bisq which do not have decent broadband internet access

### and finally
I am fully aware that there is a lot of technical dept in these parts of the code. And the dept needs to go. But we cannot eliminate this dept in one step (well, actually, we might be able to pull it off, but that would be a complete rewrite of the p2p module and all the risks that comes with touching this part of the code). And it seems like we are discussing stuff over and over again, not coming to a consensus for months. Meanwhile, the actual request size has grown beyond 6MB by now and we are denying more and more users access to the bisq network (and thereby loosing them).

The actual bad thing here is that the code stayed as is, no technical dept added, non removed, nothing changes. Just a massive waste of time.

freimair committed:
Included PR feedback

freimair committed:
Included PR feedback

ripcurlx committed:
Bump version number for v1.3.8

ripcurlx committed:
Update translations for v1.3.8

chimp1984 committed:
Add dontShowAgain button in osxKeyLoggerWarning window

chimp1984 committed:
Rmove short cut info for removed key handlers

chimp1984 committed:
Refactor dependency structure to enable adding dispute managers

If we would add DisputeManager to previous structure it would cause a
circular dependency error from guice. We change dependency structure so
that TradeManager does not know XmrTxProofService but XmrTxProofService
gets an instance of TradeManager. It makes code cleaner in total as well
as responsibility is better defined.

Next commit will contain the DisputeManager addition.

chimp1984 committed:
Add check if dispute is open

If a mediation or arbitration dispute has been opened we do not use the
auto-confirm feature.

chimp1984 committed:
Enable onFiatPaymentReceived again (was for dev testing)
- Remove commented out code
- do isFiatReceived in stream filter

chimp1984 committed:
Ensure that p2p network and wallet are ready

chimp1984 committed:
Make var local

chimp1984 committed:
Set default services based on localhost check not on devMode check

chimp1984 committed:
Change min required confirmations from 0 to 1.

If users accept visible in mempool only txs its their own risk and they
can manually confirm anyway. We should not support 0 conf txs.

chimp1984 committed:
Add monero orange color code for tx confidence indicator

chimp1984 committed:
Show tx confidence indicator for XMR blocks. Show info icon

chimp1984 committed:
Adjust auto-conf css for badge

- make it xmr orange
- make with dynamic with 10 px padding left/right
- fix 1 px vertical offset

chimp1984 committed:
Add null check. Improve text

chimp1984 committed:
Do not overwrite useDevMode with useDevModeHeader

If --useDevModeHeader is not set it is false by default.
If user has --useDevMode=true set it would overwrite his value.

chimp1984 committed:
Change hex color code to lower case

chimp1984 committed:
Fix bug with all mobile notifications for disputes are sent at startup.

chimp1984 committed:
Improve handling of p2pNetworkAndWalletReady

The p2pNetworkAndWalletReady MonadicBinding might be removed from GC
if its a local variable. I observed that in BisqSetup with a similar
setup. It might be an implementation weakness in MonadicBinding
(usage of weak references?). A tester reported that he does not see any
result, which might be cause that the service never gets the
onP2pNetworkAndWalletReady triggered if the MonadicBinding is not there
anymore.
By removing the listener we need at shutdown we need it anyway as class
field (so codacy does not complain anymore). As well added a check if
all is already complete to skip the MonadicBinding at all
(not expected case in onAllServicesInitialized).

chimp1984 committed:
Move remove code from shutDown to onP2pNetworkAndWalletReady

chimp1984 committed:
Add JsonExclude to userName

For backward compatibility we need to exclude the new field for
the contract json.
We can remove that after a while when risk that users with
pre 1.3.8 version trade with updated
users is very low.

chimp1984 committed:
Add check if xmrTxProof is used to avoid null pointer exception

chimp1984 committed:
Move revolutAccountsUpdateHandler code up to BisqSetup

chimp1984 committed:
Fix handling of username/accountId

We apply userName to accountId if it is not set (e.g. new account
created with new version). We do not use that for display or for
account signing in case both fields are the same but we need to use
accountId in case the user trades with a not updated user who expects
accountId as only field.
I improved a bit the display of account data in the trade screens.
In case accountId was set with the phone number (updated account with
phone nr used for account signing) we show both userName and phone nr.

- Show phone number if accountId was set by old account. Otherwise
show only userName
- For old users they will see the user name as phone number displayed
if they trade with new users if the new user has created a new account.
If he has updated an existing account the accountId (phone number) is
used, so it displays the phone number.
- At step 2 changed display of own account data to show account name
- Add 'Recipients' prefix to account data of peer at step 2

Step 3: Buyers account data can be
- Phone number if peer is using old version
- User name if peer is updated user with new account (we apply userName
to accountId)
- Phone number if user is on old version and peer is updated user with
updated account (we keep accountId as phone number)
- User name/Phone number if peer is updated user with updated account

chimp1984 committed:
Scan disputes for accounts where same user used diff. real names.
Might be fraudulent traders.

chimp1984 committed:
Remove unused var

chimp1984 committed:
Add alert icon to list entries

Support agent can mark a suspicious dispute as resolved so it does not
show the alert icon anymore. In the full report a [ACK] got added to
that dispute.

chimp1984 committed:
Remove unused method

chimp1984 committed:
Resolve merge conflict in Merge conflict in desktop/src/main/java/bisq/desktop/main/portfolio/pendingtrades/steps/seller/SellerStep3View.java

chimp1984 committed:
Resolve merge conflict in Merge conflict in desktop/src/main/java/bisq/desktop/main/portfolio/pendingtrades/steps/seller/SellerStep3View.java

chimp1984 committed:
Remove offer from takers offerbook after take offer is completed.

We do not wait until the offer got removed by a network remove message but remove it
directly from the offer book. The broadcast gets now bundled and has 2 sec. delay so the
removal from the network is a bit slower as it has been before. To avoid that the taker gets
confused to see the same offer still in the offerbook we remove it manually. This removal has
only local effect. Other trader might see the offer for a few seconds
still (but cannot take it).

wiz committed:
Add new Monero Explorer node77.monero.wiz.biz with onion

wiz committed:
Fix compile error in core/src/main/java/bisq/core/user/Preferences.java

chimp1984 committed:
Improve handling of nullable assetTxProof fields

Commit:
Minor bug fixes for XMR tx proof feature

Fix missing CSS color code xmr-orange, was missing from dark mode.

Fix log message spelling/typo errors.

Removed 2 fixes from SellerStep3View so that chimp1984 can make
changes.

Remove address validator from XMR service address settings because
it does not support https prefix.

chimp1984 committed:
Bug fixes

- useXmrTxProof was set at active but used in addContent which was
called before activate.
- Fix bug with using currencyName instead of currencyCode
- Refactor: Use methods in base class instead of repeated code

chimp1984 committed:
Add comment to empty method

chimp1984 committed:
Add check if trade have been dispute to protocol classes

chimp1984 committed:
Use a checkArgument to ensure that the methods are not called once a
dispute has been opened. This will cause a Runtime exception but that
is justified as the caller need to ensure to do the check and do not
allow to get to that point.

chimp1984 committed:
Resolve merge conflict in Merge conflict in desktop/src/main/java/bisq/desktop/main/portfolio/pendingtrades/steps/seller/SellerStep3View.java

chimp1984 committed:
Create new HttpClient for each request.

chimp1984 committed:
Resolve conflict in 6a061d2c5

CONFLICT (modify/delete): core/src/main/java/bisq/core/trade/txproof/xmr/DevTestXmrTxProofHttpClient.java

chimp1984 committed:
Remove DevTestXmrTxProofHttpClient

chimp1984 committed:
Remove DevTestXmrTxProofHttpClient

wiz committed:
Use lots of regex to add http or https for XMR explorer API endpoint

* If Tor *.onion hostname, use HTTP with Tor proxy
* If 127.0.0.1 or localhost, use HTTP without Tor proxy
* If LAN address or *.local FQDN, use HTTP without Tor proxy
* If any other FQDN hostname, use HTTPS with Tor proxy

Emzy committed:
Change from http port 80 clearnet to https 443

To be compatible to https://github.com/bisq-network/bisq/pull/4492

Commit:
Add note to SetXmrTxKeyWindow

To convey what it's about, and that it's optional.

wiz committed:
Tweak string for "Service Addresses" to "Monero Explorer URLs"

wiz committed:
Replace old Bisq Explorer URLs with new Mempool URLs for the 3 we have

wiz committed:
Tweak strings of "BSQ Explorer" -> "Bisq Explorer" and "Bitcoin Explorer"

wiz committed:
Fix codacy complaining about missing space character

wiz committed:
Rename Monero Explorer to friendlier hostname explorer.monero.wiz.biz

chimp1984 committed:
Ignore future results in case we had a terminal result already.

Avoid that a success result overwrites an earlier failed/error result.

chimp1984 committed:
Rearrange fields and improve toString method

chimp1984 committed:
Improve text in UI in case of failure or error

chimp1984 committed:
Remove side effect in setUserName method and add extra handling for the
moment we save the account. Only at that moment we check if we need to
set the accountId with the value of the userName.
We do that in the domain layer to avoid more domain logic code in the UI
layer.

Fixes bug found at:
https://github.com/bisq-network/bisq/pull/4481#pullrequestreview-485066342

ripcurlx committed:
Update data stores for v1.3.8

devinbileck committed:
Add new Monero explorer xmrblocks.bisq.services with onion

As per https://github.com/bisq-network/proposals/issues/257

wiz committed:
Revert Tether USD for now due to various issues

freimair committed:
Added 1.3.8 data stores

freimair committed:
Use Version.VERSION again

I thought I could get by using a fixed version string so we can
have tests involing future data stores...

chimp1984 reviewed (changes requested):
NACK

Basic concept is ok but many impl. elements are not acceptable IMO. Specially the hack to encode the version into the payload of the hashes instead of adding a new protobug field to the request msg.

I suggest that we use #4519 which is based on this PR one but has a lot of changes. It is still WIP and not much tested, should be ready for review in a few days...

chimp1984 commented:
@sqrrm @ripcurlx Can you close that as it was replaced by a merged PR
